nav2_dwb_controller/dwb_plugins/src/kinematic_parameters.cpp

```diff
@@ -117,12 +129,41 @@ void KinematicsHandler::initialize(
   parameter_event_sub_ = parameters_client_->on_parameter_event(
     std::bind(&KinematicsHandler::on_parameter_event_callback, this, _1));
 
+  if (kinematics.use_speed_limit_) {
+    max_vel_sub_ = nh->create_subscription<nav2_msgs::msg::SpeedLimit>(
+      kinematics.speed_limit_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+      std::bind(&KinematicsHandler::speedLimitCallback, this, std::placeholders::_1));
+  }
+
   kinematics.min_speed_xy_sq_ = kinematics.min_speed_xy_ * kinematics.min_speed_xy_;
   kinematics.max_speed_xy_sq_ = kinematics.max_speed_xy_ * kinematics.max_speed_xy_;
 
   update_kinematics(kinematics);
 }
 
+void KinematicsHandler::speedLimitCallback(
```

Comment:
I thought we had discussed moving the topic speed limit handling to the controller server, that way every single controller plugin doesn't need to implement this. If we have the controller server handle it, then we can just update the controller plugin API to have a `setSpeedLimit()` method that will override its maximum set speed.

Reply:
Let's confirm that we are on the same page:
- Subscription to the speed limiting topic will be in the `ControllerServer`
- Also, there are `use_speed_limit: True/False` and `speed_limit_topic: topic_name` ROS-parameters will be in `ControllerServer`
- Adding new `setSpeedLimit()` API method to:
1. `Controller` basic interface
2. `DWBLocalPlanner` derivative plugin implementation as `: public Controller`
3. `TrajectoryGenerator` abstract class ([part](https://github.com/ros-planning/navigation2/tree/main/nav2_dwb_controller#changes-to-dwb) of DWB)
4. `StandardTrajectoryGenerator` derivative plugin implementation as `: public TrajectoryGenerator`
5. `KinematicsHandler` as "key" function that actually changes the speed

This scheme makes speed limiting to be unified for all controllers and plugins, but pretty complicated in my opinion. Could you please confirm the design of change?

Reply:
I don't think the `use_speed_limit` is needed. If they didn't want to use it, they should just not use the `speed layer` which even sets it. Yes on the topic, though the default should match the layer's defaults so that they don't need to play with it for the typical usecase. I think that also means that this speed layer needs to only be run in **either** the global **or** local costmap layer. We should make sure that's immediately clear in docs / descriptions. 

`void setSpeedLimit(const double & percentage)` or `void setSpeedLimit(const double & abs_velocity)`? That also begs a question if its `abs_velocity` how we represent it for non-differential robots (e.g. can move in Y or even Z directions). Maybe the speed layer should only support percentages... 

And yes, in the main controller `nav2_core` header that all should implement. I haven't thought out the specific implementation of DWB, but updating the kinematics is the ultimate goal. I'm not sure off hand who owns that object to know if the generators need to be involved.

Reply:
> I don't think the use_speed_limit is needed. If they didn't want to use it, they should just not use the speed layer which even sets it.

The `use_speed_limit` is not strictly required for other developers. This is rather made for optimization in order to do not add one more subscription every time even if we are not going to use it.

> I think that also means that this speed layer needs to only be run in either the global or local costmap layer. We should make sure that's immediately clear in docs / descriptions.

Agree with it. I will note this in docs for speed filter tutorial.

> void setSpeedLimit(const double & percentage) or void setSpeedLimit(const double & abs_velocity)? 

Why not `void setSpeedLimit(const nav2_msgs::msg::SpeedLimit & speed_limit)`?

> That also begs a question if its abs_velocity how we represent it for non-differential robots (e.g. can move in Y or even Z directions). Maybe the speed layer should only support percentages...

Really good question. I think, we need to restrict the absolute/percentage value of linear velocity vector (be it casual car-like robot, differential drive, omni-directional drive or drone moving in 3D space):
![speed_filter_velocity](https://user-images.githubusercontent.com/60094858/99414299-bb5ee380-2907-11eb-854a-694745b3c976.jpg)
It is up to controller plugin how to handle it.

For DWB the situation is quite tricky: both velocity components: `vel_x` and `vel_y` are initially handled independently and then velocity is being finally corrected by its absolute value. For independent handles there are `max_vel_x` and `max_vel_y` parameters are responsible, for last correction - `max_speed_xy`. Theoretically, we can limit only `max_speed_xy` parameter in kinematics. As I've checked today, this will be enough to work. But just in case, lets restrict  `max_vel_x` and  `max_vel_y` if the given speed limit is less than current component value. The code snap of algorithm might be as follows:
```
  if (msg.speed_limit == nav2_costmap_2d::NO_SPEED_LIMIT) {
    // Restore default value
    kinematics.max_vel_x_ = kinematics.max_vel_x_default_;
    kinematics.max_vel_y_ = kinematics.max_vel_y_default_;
    kinematics.max_speed_xy_ = kinematics.max_speed_xy_default_;
  } else {
    if (msg.percentage) {
      // Speed limit is expressed in % from maximum speed of robot
      kinematics.max_vel_x_ = kinematics.max_vel_x_default_ * msg.speed_limit;
      kinematics.max_vel_y_ = kinematics.max_vel_y_default_ * msg.speed_limit;
      kinematics.max_speed_xy_ = kinematics.max_speed_xy_default_ * msg.speed_limit;
    } else {
      // Speed limit is expressed in absolute value
      if ( kinematics.max_vel_x_ > msg.speed_limit) {kinematics.max_vel_x_ = msg.speed_limit; }
      if ( kinematics.max_vel_y_ > msg.speed_limit) {kinematics.max_vel_y_ = msg.speed_limit; }
      kinematics.max_speed_xy_ = msg.speed_limit;
    }
  }
```
This will work both for absolute and percent speed values in speed_filter. So, I see any reason to drop absolute speed limits support. How do you think?

> Perhaps have the generator give the main DWB local planner pointer access to it to update.

Unfortunately. As I've found, `DWBLocalPlanner` has a `StandardTrajectoryGenerator` plugin inside which has shared pointer to `KinematicsHandler` being protected. So, we need to add the chain I've described in prev. comment or either add `getKinematicsHandler()` new method. First way seems better for me, since I fell `KinematicsHandler` class is optional and might vary from implementation of `TrajectoryGenerator` plugin.

>  If they don't want to support the speed limit layer, they could always just have this function do nothing.

Thus, it is better to force developers to implement their own API routine, say with following intention in the basic `Controller` class:
```
virtual void setSpeedLimit(const nav2_msgs::msg::SpeedLimit &) = 0;
```

Reply:
> The use_speed_limit parameter is not strictly targeted for developers. This is rather made for optimization in order to do not add one more subscription every time even if we are not going to use speed filter. But I am OK to remove this parameter at all.

Good point. I leave it up to you, I would accept either way. I generally try to avoid derived parameters we can get from other properties since Nav2 already has like 400 parameters.

> Why not void setSpeedLimit(const nav2_msgs::msg::SpeedLimit & speed_limit)?

I suppose my goal is to keep ROS interfaces limits to where they're minimally required. I looked over the message and it looks like just a `double limit, bool is_percentage`. Maybe we just change the API to include the `bool`. But it fed into the next question on how to deal with absolute velocity values with non-differential robots. If we do only percentage, things simplify a little. We'll cover that separate discussion below.

> We need to restrict the absolute/percentage value of linear velocity vector

That still doesn't resolve it. If we set the absolute velocity to 1, that doesn't mean `1` should apply to X and Y. I could see an omni robot having a fast X speed but a slower Y since that's mostly used when making fine maneuvers into final pose. Or Z on a drone the climb rate is decoupled from the planar motion. If anything when using absolute speeds, the value would have to be the length of the velocity vector (for omni the `xy` speed, diff still just `x`) but doesn't really answer how the angular velocities for even differential drive robots should react. 

I think that makes we actually can only do percentage on this without the mask containing more than 1 value per pixel. Even for the most basic case of differential robots, we need 2 values, for X and theta. We wouldn't want to drop the speed to 50% speed via absolute parameters, but then still allow the robot to rotate and whip around at full speed (especially if non-circular). 

For the angular velocity reason alone, I think percentage is the only way this works. I think we could work out some compromise for the linear velocities, but this is a hole that we didn't think about until now :-(   

> getKinematicsHandler() ...

I think `getKinematics()` is cleaner, but I agree with that concept. You mention that it will vary by implementation of trajectory generator plugin, but wouldn't that be true of the 'chain' method you introduced as well? I looked it over, while the standard trajectory generator is the only one that has an instance of the kinematics handler, the limited acceleration generator is derived from the standard trajectory generator so it also uses it. The question is then if we decide to include `N` new generators, do we want to enforce that they all use this handler object. 

I think that answer is tentatively "yes" since it handles all the major parameters and the callback updates https://github.com/ros-planning/navigation2/blob/main/nav2_dwb_controller/dwb_plugins/src/kinematic_parameters.cpp and it would be convenient to share this across all of them. However I do see a gap where `angle` params would have to be misused to be `turning_angle` for an ackermann trajectory generator, which is in my (hopefully) start of 2021 tasks (unless someone beats me to it). So the question is then, is there a way to modify kinematics handler to be more configurable? I think that ultimately is the question that will tell us if we can do something like `getKinematics()` or if we need to chain it own. 

> Thus, it is better to force developers to implement their own API routine (even empty)...

Exactly :-)

Reply:
Hm... so the main question is in maximum angular speed does not changed when we specifying linear speed limit in _absolute value_?

How about such "workaround":

- When linear velocity is arriving to a controller plugin, we calculating the `relation = msg.speed_limit / max_linear_speed_xyz_`. And then applying it to twist: `max_angular_speed_ = max_default_angular_speed_ * relation`. This will slowing down angular velocity of the robot in the same proportion as it will be slowed down by linear velocity value.

This is not clean solution, rather a dirty workaround :-D But this might be a compromise for the linear velocities you've mentioned about...

Reply:
Yeahhhh I'm not really a fan of that, because its not then really representing the values but a "hack" that has to be implemented in each specific plugin individually to enable. It ends up essentially being a percentage scale for all other values.

I agree that having an option for absolute velocity would be nice, but I think there's some features here we can't do because we're trading for flexibility in arbitrary robots. 

Maybe there's another solution to handle absolute velocities? I think the major reason we discussed having both is in the case of heterogeneous fleets making sure that `N` robot manufacturers with different max speeds can respect an absolute speed limit in certain areas. I'm not sure right now how we could support that cleanly. Another idea could be to have a separate costmap filter layer for "safety speed limits" where it takes in a binary mask and for the non-0 values it applies a parameterized maximum speed. That way all robots with this later are binary "in" a safety speed zone to be restricted to some absolute speed, or not and use max speed. That wouldn't really play well with having both these layers in use at the same time however. Supporting both percentage and absolute in the same layer I think needs some new ideas.

Reply:
> I agree that having an option for absolute velocity would be nice, but I think there's some features here we can't do because we're trading for flexibility in arbitrary robots.

That is why it is not a clean solution. To be honest, we need two masks for one filter: one mask with encoded absolute max linear velocity, another - with absolute max. angular (twist) velocity. But I am not a fan of such solution, at least because the system setup will be really heavy to understand and use by other developers.

> Another idea could be to have a separate costmap filter layer for "safety speed limits" where it takes in a binary mask and for the non-0 values it applies a parameterized maximum speed. That way all robots with this later are binary "in" a safety speed zone to be restricted to some absolute speed, or not and use max speed.

In current design it works exactly like you've described. Using one mask is enough for absolute values: "0" values on map - are free areas (no speed limiting). Non-zero data (`data=[1..100]` in `OccupancyGrid`) means that marked area is speed limited. Absolute speed (linear velocity) limit will be decoded as `msg.speed_limit = base + data * multiplier`. This speed will be a speed limit, if it less than default maximum speed for the robot.

> That wouldn't really play well with having both these layers in use at the same time however. Supporting both percentage and absolute in the same layer I think needs some new ideas for encoding information.

Hm... it looks like here I start to lose the thread of reasoning. What a use-case of two speed filters (1: percentage, 2: absolute) acting at the same time?

Reply:
> Maybe there's another solution to handle absolute velocities?

Running two `speed_filter` plugins at the same time: one tuned for publishing maximum linear velocity, another - for publishing maximum twist velocity. Also, required update the field in "nav2_msgs::msg::SpeedLimit" message:
```
std_msgs/Header header
# Setting speed limit in percentage if true or in absolute values in false case
# type=0 - percent speed limit
# type=1 - absolute max. linear velocity limit
# type=2 - absolute max. angular velocity limit
uint8t type
# Maximum allowed speed (in percent of maximum robot speed or in m/s depending
# on "percentage" value). When no-limit it is set to 0.0
float64 speed_limit
```
in order to allow controller to define which component of speed or percentage it is working with.

Reply:
My thought process at this point is to potentially drop the absolute speed feature altogether and focus on the percentage feature since that's better scaled and described in a single value and gives expected outputs. I'd be fine discussing ideas how to cleanly enable absolute speeds later, but I think we need to go back to the drawing board a little bit in light of these new issues for describing angular velocities and other non-forward-X velocities.  

The linear scaling of having the value be X absolute velocity and then scaling the other directions and angular velocities by its scale seems reasonable. I just wonder though if that's "enough" to meet the needs that it was setting out to fulfill. I think we need to bring this up to a working group meeting and see what folks think on that topic.

Reply:
I am not very happy about it, but we better to move absolute values to be treated separately in another ticket. So, if we will drop abs. values support, what will be about `nav2_msgs/msg/SpeedLimit.msg`. Is it reasonable to remain there only 1 field for now?
```
$ cat nav2_msgs/msg/SpeedLimit.msg
std_msgs/Header header
# Maximum allowed speed (in percent of maximum robot speed).
# When no-limit it is set to 0.0
float64 speed_limit
```

Reply:
I think leaving the message as is is preferable. Its still a useful message that should have flexibility even if no application uses it yet. Someone else could use the same message for a different application .

Reply:
Temporarily removed the support of absolute speed values in https://github.com/ros-planning/navigation2/commit/400c1200ec5967b34a1e6e9907ae79e1326faaae commit. I think, after discussion with the Community, we can make separate PR with absolute speeds support adding.